File: 백준/Gold/1052. 물병/물병.py
import sys

n, k = map(int, sys.stdin.readline().split())
binary = []
num_bottles = 0
while n > 0:
    binary.append(n & 1)
    num_bottles += n & 1
    n >>= 1

answer = 0
buying = False
for i in range(len(binary)):
    if buying:
        if binary[i] == 0:
            binary[i] = 1
            buying = False
        else:
            num_bottles -= 1
    
    if num_bottles <= k: break
    
    if not buying and binary[i] == 1:
        buying = True
        answer += 2**i

print(answer)

File: 백준/Gold/1052. 물병/물병.cc
#include <iostream>
#include <string>
#include <vector>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);

using namespace std;

int main() {
    FASTIO;
    int n, k;
    cin >> n >> k;

    vector<int> binary;
    int num_bottles = 0;
    while(n > 0){
        binary.push_back(n & 1);
        num_bottles += n & 1;
        n >>= 1;
    }

    int answer = 0;
    bool buying = false;
    for(int i = 0; i < binary.size(); i++){
        if(buying){
            if(binary[i] == 0){
                binary[i] = 1;
                buying = false;
            }
            else num_bottles--;
        }

        if(num_bottles <= k) break;

        if(!buying && binary[i] == 1){
            buying = true;
            answer += 1 << i;
        }
    }

    cout << answer << endl;
    
    return 0;
}